Add Upsells collection

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds a new Upsells collection to Product Collection block.

Closes https://github.com/woocommerce/woocommerce/issues/50860

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites:
1. Make sure you have:
    - at least 2 products with some Upsells
    - at least 1 product with no Upsells.
Here are instructions [on how to set them up](https://woocommerce.com/document/related-products-up-sells-and-cross-sells).

#### Single Product template
1. Go to Editor -> Single Product template
3. Add Product Collection block
4. Choose Upsells collection
    - Expected: it previews 8 products
5. Save and go to frontend
6. Visit product with upsells:
     - Expected: there's collection rendered with "You may also like" heading and upsells products
7. Visit product with no upsells:
    - Expected: collection is not rendered at all (no heading as well)

Example: 
Product editor | Frontend
-- | --
<img width="608" alt="image" src="https://github.com/user-attachments/assets/a4f13dd9-b329-4ddb-a3a6-d0a46f511c57"> | <img width="1117" alt="image" src="https://github.com/user-attachments/assets/a28e298c-7bd7-4e7a-b045-7adb2762fb47">

#### Specific product template

1. Go to Editor -> Add new template
2. Choose "Single item: Product"
3. Choose a product that has Upsells
5. If possible choose suggested template, it should already include Upsells. If not, just start blank and add Product Collection and choose Upsells
6. Expected: Actual Upsells for this product are displayed in Editor

<img width="1097" alt="image" src="https://github.com/user-attachments/assets/2f659491-e2ab-4490-b841-7988cf3c1406">

#### Cart

1. Go to Editor -> Page: Cart
2. Add Product Collection on the page
3. Choose Upsells
    - Expected: it previews 8 products
5. Save and go to frontend
7. Add to cart a product with NO upsells
8. Go to cart
    - Expected: collection is not rendered at all (no heading as well)
9. Add to cart a product with upsells
     - Expected: there's collection rendered with "You may also like" heading and upsells products
10. Add another product with upsells
     - Expected: upsells from two products are merged (no specific order). There's no duplicates!
11. Add one of the upsell products to cart. Refresh the page
     - Expected: the product you added to cart no longer appears in upsells.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ObliviousHarmony committed:
Fixed PHP 7.4 Compatibility

The splat operator does not support associative arrays until PHP 8.1.

ObliviousHarmony committed:
Added Related Products Collection Boilerplate

ObliviousHarmony committed:
Added `woocommerceRelatedTo` Collection Parameter

This parameter allows the block to query for products that are related
to the given product IDs.

ObliviousHarmony committed:
Added Collection Related Product Filter

Using `woocommerceRelatedTo`, queries may restrict the products returned
to those that are related to the given product ID(s).

ObliviousHarmony committed:
Removed Unnecessary Tracking

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Removing Misunderstood Option

ObliviousHarmony committed:
Linting Fix

ObliviousHarmony committed:
Fixed Test

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/49335-related-products-collection

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Removed Unnecessary Keyword

ObliviousHarmony committed:
Support Unlimited `wc_get_related_products()`

This changes related product fetching so that `-1` will return all
related products. I also removed a second +10 offset that seems
to have been accidentally added eight years ago.

ObliviousHarmony committed:
Removed `woocommerceRelatedTo` Param

Since we're going to use the collection name, we don't need
this functionality any longer.

ObliviousHarmony committed:
Simplified `post__in` Filtering

Since `merge_queries` was already using an intersection merge for
`post__in`, this second function is unnecessary. I've removed it
as well as refactored the merging logic.

ObliviousHarmony committed:
Pass Collection Name To Final Query

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Added Custom Collection Handlers

Developers can register collections along with handlers that implement
the custom behavior.

ObliviousHarmony committed:
Removed Second Test Query Construction

As it was, `build_query_vars_from_query_block()` was triggering the
ProductCollection instance hooked into WordPress. After that, we
called `build_frontend_query()` on our test instance. This caused some
weird behavior in tests. As a result, however, the tax_query merge
test couldn't rely on the tax_query transformation done by WordPress.

ObliviousHarmony committed:
Added Related Product Collection Handlers

ObliviousHarmony committed:
Revert "Support Unlimited `wc_get_related_products()`"

This reverts commit 41c837202ebad27031b82f4227cba0b307bc7905.

ObliviousHarmony committed:
Fixed Preview Mode Query

It looks like we were checking the wrong place for the preview
state.

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Fixed Test

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Updated Collection Heading

kmanijak committed:
Add Upsells collection definition

kmanijak committed:
Register Upsells

kmanijak committed:
Add collection handlers for Upsells

kmanijak committed:
Add changelog

kmanijak committed:
Remove on sale query param

ObliviousHarmony commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1862,5 +1862,55 @@ function ( $collection_args, $query, $request ) {
+				}
+
+				$product = wc_get_product( $collection_args['upsellsProductReference'] );
+				$upsells = array_map(
+					function ( $id ) {
+						return strval( $id );
+					},
+					$product->get_upsell_ids()
+				);
```
```suggestion
				$upsells = array_map( 'strval', $product->get_upsell_ids() );
```

Interestingly, PHP actually lets you just use the string name of the function as the argument!

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Better Collection Test Teardown

Passing a callable back means that we can use variables in the
setUp and then the tearDown without any confusion.

ObliviousHarmony committed:
Merge branch 'add/49335-related-products-collection' into add/50860-upsells

ObliviousHarmony committed:
Removed Unnecessary Test Provider

ObliviousHarmony committed:
Merge branch 'add/49335-related-products-collection' into add/50860-upsells

ObliviousHarmony committed:
Added Upsell Collection Unit Test

ObliviousHarmony committed:
Product ID Confidence Check

ObliviousHarmony committed:
Added Collection Product Reference

kmanijak committed:
Don't display Upsells collection if there's no upsells

kmanijak committed:
Change the Upsells collection heading

ObliviousHarmony reviewed (dismissed):
Thanks @kmanijak, it's great to see an implementation of the collection handlers 😄 This tests well for me, however, I'm _not_ going to approve it yet until a decision is made on the handlers. After all, this is targeting my branch!

kmanijak committed:
Add Upsells supporting cart context

kmanijak committed:
Consume the cart productIds and display Upsells in there

ObliviousHarmony committed:
Fixed `$post__in` Merging

When the intersection is empty it was returning all products. This
makes it so that it returns nothing when there's no results.

kmanijak committed:
Fix lint

kmanijak committed:
Extract the variable

kmanijak committed:
Fix lint

kmanijak committed:
Merge branch 'add/49335-related-products-collection' into add/50860-upsells

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1874,5 +1874,75 @@ function ( $collection_args, $query, $request ) {
+			'woocommerce/product-collection/upsells',
+			function ( $collection_args ) {
+				$product_reference = $collection_args['upsellsProductReferences'] ?? null;
+				// No products should be shown if no upsells product reference is set.
+				if ( empty( $product_reference ) ) {
+					return array(
+						'post__in' => array( -1 ),
+					);
+				}
```
Excellent work, Karol! I really appreciate how you've handled this edge case 💪🏻 This is important because there will be cases when the merchant might not have selected a product reference.

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1874,5 +1874,75 @@ function ( $collection_args, $query, $request ) {
+					[]
+				);
+
+				// Remove duplicates and product references. We don't want to display
+				// what's already in cart.
+				$unique_upsells = array_unique( $all_upsells );
+				$upsells        = array_diff( $unique_upsells, $product_reference );
```
💜

imanish003 reviewed (approved):
Hey @kmanijak,

I tested the changes and everything seems to be working as expected. Here is what I tested:

**Single Product template**
- ✅ It's showing 8 products in the editor on Single Product template
- ✅ Product with Upsells:
	- ✅ Rendered with "You may also like" heading
	- ✅ Correct Upsell products are showing on frontend
- ✅ Product without Upsells:
	- ✅ Not rendered

**Specific product template**
- ✅ Upsells are showing in the editor on Specific Product template

**Cart**
- ✅ On Editor, it previews 8 products in preview mode
- ✅ Upsells collection is empty when product with no upsells is added to cart
- ✅ Upsells collection is rendered when product with upsells is added to cart
- ✅ Upsells from two products are merged (no specific order). There's no duplicates.

---

Code also looks good to me 🚀 I can see that there is some duplication of code between "Related Products" and "Upsells" collection but I believe refactoring can be done as part of new task.

imanish003 committed:
Resolve merge conflicts
